### PR TITLE
Fix NPEs

### PR DESCRIPTION
**Pull request by simon33-2**
Fix some NPEs caused by the last change.

### CODE REVIEW

File: src/main/java/games/strategy/triplea/delegate/DependentBattle.java [hunk 1]
```diff
@@ -18,13 +18,16 @@
...
+  protected Map<Territory, Collection<Unit>> m_attackingFromMap;
+  protected Set<Territory> m_attackingFrom;
+  private Collection<Territory> m_amphibiousAttackFrom;
 
   DependentBattle(final Territory battleSite, final PlayerID attacker, final BattleTracker battleTracker,
       final boolean isBombingRun, final BattleType battleType, final GameData data) {
     super(battleSite, attacker, battleTracker, isBombingRun, battleType, data);
+    m_attackingFromMap = new HashMap<>();
+    m_attackingFrom = new HashSet<>();
+    m_amphibiousAttackFrom = new ArrayList<>();
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/triplea/delegate/DependentBattle.java]: Are you sure this change makes sense?

Comment (simon33-2) [inline src/main/java/games/strategy/triplea/delegate/DependentBattle.java]: Well putting in the declaration doesn't work.

Comment (RoiEXLab) [inline src/main/java/games/strategy/triplea/delegate/DependentBattle.java]: This is weird, because it should...

**Comment by ssoloff** [comment]
@simon33-2 Would you happen to have a stack trace from the NPEs you encountered?  It might help put these changes into context.

**Comment by simon33-2** [comment]
I can reproduce it by running from master. Here's one on m_amphibiousAttackFrom:

triplea.engine.version.bin:1.9
Loading map: world_war_ii_global, from: /ubuntu/home/simon/src/world_war_ii_global/map
Loading resources from the following paths: [/ubuntu/home/simon/src/world_war_ii_global/map, /ubuntu/home/simon/src/triplea/assets]
Loading map: world_war_ii_global, from: /ubuntu/home/simon/src/world_war_ii_global/map
Loading resources from the following paths: [/ubuntu/home/simon/src/world_war_ii_global/map, /ubuntu/home/simon/src/triplea/assets]
java.lang.NullPointerException
	at games.strategy.triplea.delegate.BattleDelegate.doScrambling(BattleDelegate.java:686)
	at games.strategy.triplea.delegate.BattleDelegate.start(BattleDelegate.java:88)
	at games.strategy.engine.framework.ServerGame.startStep(ServerGame.java:513)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:420)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:272)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInNewThread(LocalLauncher.java:61)
	at games.strategy.engine.framework.startup.launcher.AbstractLauncher.lambda$launch$0(AbstractLauncher.java:43)
	at java.lang.Thread.run(Thread.java:745)


Here's one on one of the other two variables:
Loading map: world_war_ii_global, from: /ubuntu/home/simon/src/world_war_ii_global/map
Loading resources from the following paths: [/ubuntu/home/simon/src/world_war_ii_global/map, /ubuntu/home/simon/src/triplea/assets]
Loading map: world_war_ii_global, from: /ubuntu/home/simon/src/world_war_ii_global/map
Loading resources from the following paths: [/ubuntu/home/simon/src/world_war_ii_global/map, /ubuntu/home/simon/src/triplea/assets]
java.lang.NullPointerException
	at games.strategy.util.Match.getMatches(Match.java:35)
	at games.strategy.triplea.delegate.MustFightBattle.getAttackerRetreatTerritories(MustFightBattle.java:1218)
	at games.strategy.triplea.delegate.MustFightBattle.canAttackerRetreat(MustFightBattle.java:1258)
	at games.strategy.triplea.delegate.MustFightBattle.canAttackerRetreatSubs(MustFightBattle.java:1276)
	at games.strategy.triplea.delegate.MustFightBattle.access$1700(MustFightBattle.java:50)
	at games.strategy.triplea.delegate.MustFightBattle$14.execute(MustFightBattle.java:889)
	at games.strategy.triplea.delegate.ExecutionStack.execute(ExecutionStack.java:35)
	at games.strategy.triplea.delegate.MustFightBattle.fight(MustFightBattle.java:326)
	at games.strategy.triplea.delegate.BattleDelegate.fightBattle(BattleDelegate.java:227)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.delegate.DelegateExecutionManager$1.invoke(DelegateExecutionManager.java:148)
	at com.sun.proxy.$Proxy9.fightBattle(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:172)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:150)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:132)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWait(UnifiedMessenger.java:101)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:50)
	at com.sun.proxy.$Proxy9.fightBattle(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.gamePlayer.GameOverInvocationHandler.invoke(DefaultPlayerBridge.java:162)
	at com.sun.proxy.$Proxy9.fightBattle(Unknown Source)
	at games.strategy.triplea.TripleAPlayer.battle(TripleAPlayer.java:484)
	at games.strategy.triplea.TripleAPlayer.start(TripleAPlayer.java:131)
	at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:531)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:424)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:259)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInNewThread(LocalLauncher.java:61)
	at games.strategy.engine.framework.startup.launcher.AbstractLauncher.lambda$launch$0(AbstractLauncher.java:43)
	at java.lang.Thread.run(Thread.java:745)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:62)
	at com.sun.proxy.$Proxy9.fightBattle(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.gamePlayer.GameOverInvocationHandler.invoke(DefaultPlayerBridge.java:162)
	at com.sun.proxy.$Proxy9.fightBattle(Unknown Source)
	at games.strategy.triplea.TripleAPlayer.battle(TripleAPlayer.java:484)
	at games.strategy.triplea.TripleAPlayer.start(TripleAPlayer.java:131)
	at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:531)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:424)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:259)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInNewThread(LocalLauncher.java:61)
	at games.strategy.engine.framework.startup.launcher.AbstractLauncher.lambda$launch$0(AbstractLauncher.java:43)
	at java.lang.Thread.run(Thread.java:745)


Helpful?

**Comment by RoiEXLab** [comment]
Hmm.. I believe this is a bug, but whatever...

**Comment by simon33-2** [comment]
You mean a Java bug?

**Comment by simon33-2** [comment]
Ok, I've tried downloading build 3946 on Lubuntu and on Mac OS X 10.7.5 and in both cases I can duplicate both bugs. So if it's a Java problem it affects the build or 2/3 plausible platforms.

So it needs to be fixed.

Obviously, it is retarded if an abstract class can syntactically initialise its properties' values but then doesn't use them. Should be a compiler error if it doesn't work. "Don't fix the bugs, change the documentation. We'll call them features." - Dilbert's pointy haired boss.

**Comment by ssoloff** [comment]
@simon33-2 

> Helpful?

How about "interesting?" 😃 

I see that you're using the `world_war_ii_global` map.  Can you confirm you are using the "World War II Global 1940 Original" variant?  That is the only game I can see that has `canScramble` set to `true`.

I tried to naively reproduce the NPE by having Germany amphibiously assault Novgorod.  Unfortunately, `amphibFromTerrs` on `BattleDelegate.java:686` was non-`null` and didn't raise an NPE.  Could you please provide the repro steps you used?

**Comment by simon33-2** [comment]
Try these two game saves

[NPE.zip](https://github.com/triplea-game/triplea/files/891896/NPE.zip)

I think it's Balanced mod - I think that canScramble will be true for every map in global. Grep'ing it shows so.

**Comment by ssoloff** [comment]
@simon33-2 

Thanks for the saves.  I will investigate.

> I think that canScramble will be true for every map in global. Grep'ing it shows so.

Yes, you are correct.  I was using the GitHub repo search interface, which only returned a match in `ww2global40.xml`.  Maybe GitHub search skips files larger than a certain size?  ¯\\\_(ツ)\_/¯

**Comment by ssoloff** [comment]
@simon33-2 Ok, not 100% sure on this, but it appears to be a deserialization problem.

I stepped through the deserialization process of `triplea_39594_Jap5.tsvg` and see that all the `DependentBattle` fields are `null` immediately after a `NonFightingBattle` is deserialized, while the `AbstractBattle` and `NonFightingBattle` fields appear to be ok.  (I confirmed the same for instances of `MustFightBattle` that were deserialized.)

Did you happen to create that save game prior to making the changes in #1583?  It would make sense if those fields were part of `MustFightBattle` or `NonFightingBattle` when serialized, but then they have been moved to `DependentBattle` upon deserialization.  I tried to verify this by dumping the (ungzipped) save game file, but the tool I'm using is having trouble parsing certain aspects of the serialization stream.

However, I was able to grep the (again ungzipped) save game file for class names.  I was able to locate `AbstractBattle`, `MustFightBattle`, and `NonFightingBattle`, but received no matches for `DependentBattle`.  Again, this leads me to believe the save game file was created before `DependentBattle` was introduced.

**Comment by ssoloff** [comment]
As @DanVanAtta implied [here](https://github.com/triplea-game/triplea/pull/1583#issuecomment-289189000), from a correctness perspective, I think it was necessary to change the serial version ID of both `MustFightBattle` and `NonFightingBattle` in #1583 as their class descriptors were changed in a backwards-incompatible manner (namely, the removal of fields).  I'm assuming such a change would require bumping the major version of TripleA itself?  (I don't yet know the TripleA semver rules, but I'm assuming at least some component of the version would have to be bumped.)

I was able to get the `NonFightingBattle` scramble to Sea Zone 39 to run without an NPE with the following change:

```diff
diff --git a/src/main/java/games/strategy/triplea/delegate/DependentBattle.java b/src/main/java/games/strategy/triplea/delegate/DependentBattle.java
index 85cded7..2033289 100644
--- a/src/main/java/games/strategy/triplea/delegate/DependentBattle.java
+++ b/src/main/java/games/strategy/triplea/delegate/DependentBattle.java
@@ -20,7 +20,7 @@ public abstract class DependentBattle extends AbstractBattle {
   private static final long serialVersionUID = 9119442509652443015L;
   protected Map<Territory, Collection<Unit>> m_attackingFromMap = new HashMap<>();
   protected Set<Territory> m_attackingFrom = new HashSet<>();
-  protected final Collection<Territory> m_amphibiousAttackFrom = new ArrayList<>();
+  protected Collection<Territory> m_amphibiousAttackFrom = new ArrayList<>();
 
   DependentBattle(final Territory battleSite, final PlayerID attacker, final BattleTracker battleTracker,
       final boolean isBombingRun, final BattleType battleType, final GameData data) {
diff --git a/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java b/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java
index d9bb032..86e870f 100644
--- a/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java
+++ b/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java
@@ -1,5 +1,7 @@
 package games.strategy.triplea.delegate;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
@@ -39,6 +41,13 @@ public class NonFightingBattle extends DependentBattle {
     super(battleSite, attacker, battleTracker, false, BattleType.NORMAL, data);
   }
 
+  private void readObject(final ObjectInputStream in)
+      throws IOException, ClassNotFoundException {
+    m_amphibiousAttackFrom = (Collection<Territory>) in.readObject();
+    m_attackingFrom = (Set<Territory>) in.readObject();
+    m_attackingFromMap = (Map<Territory, Collection<Unit>>) in.readObject();
+  }
+
   @Override
   public Change addAttackChange(final Route route, final Collection<Unit> units,
       final HashMap<Unit, HashSet<Unit>> targets) {
```

Presumably, a similar change could be made to `MustFightBattle`.  But this is **extremely fragile** and probably would just make the code that much harder to maintain.

**Comment by simon33-2** [comment]
I think there's two basic choices here:
1) Close this PR with unmerged commits and bump the version to 1.9.1
2) Merge this PR.

Of course, moving to 1.9.1 has issues particularly with the latest version check but these issues can be cut off in a way clear enough to the users by just telling them to migrate to 1.9.1 and including the 1.9.0 old jar.

There's other reasons to go to 1.9.1 in #1599 so I'm strongly leaning that way FWIW.

I don't think we can accept such fragile serialisation within the same version.

**Comment by simon33-2** [comment]
One more thing. If we update to 1.9.1 I think we should also update the name of the latest version file. We can't update the contents of the existing file and we also don't want to have more releases out there which can't check their latest version.

**Comment by ssoloff** [comment]
@simon33-2 I merged the changes from **this** PR into `master`, and I'm still getting the NPE at `BattleDelegate.java:686` with `triplea_39594_Jap5.tsvg`.  That kind of makes sense because constructors of serializable classes are [not called](http://stackoverflow.com/a/8141601/3900879) during the deserialization process.  For the hierarchy under consideration, that means only the `Object` default constructor will be called for an instance of `MustFightBattle` or `NonFightingBattle`; the constructors of `AbstractBattle`, `DependentBattle`, `MustFightBattle`, and `NonFightingBattle` will not be called (you can verify this by setting a breakpoint on each of these constructors and observing that they are never hit when opening your saved game).

Therefore, I'm not sure how this PR fixes the NPEs.  Maybe I'm overlooking something?  Are you doing something other than opening your save and then starting a local game?

> I think there's two basic choices here

Here's a third for your consideration:

Your refactoring in #1583 is a good change, but given the backwards compatibility requirements of serialization, you may have to approach it from a different angle.  You can still extract `DependentBattle` and consolidate the logic in `BattleDelegate`.  However, you'd have to leave the duplicate fields in `MustFightBattle` and `NonFightingBattle` and mark the `DependentBattle` accessors as `abstract`, which would require the subclasses to override them and return their private copies of the duplicate fields.

The change I'm proposing is the following:

```diff
diff --git a/src/main/java/games/strategy/triplea/delegate/DependentBattle.java b/src/main/java/games/strategy/triplea/delegate/DependentBattle.java
index 85cded7..051fe88 100644
--- a/src/main/java/games/strategy/triplea/delegate/DependentBattle.java
+++ b/src/main/java/games/strategy/triplea/delegate/DependentBattle.java
@@ -1,11 +1,7 @@
 package games.strategy.triplea.delegate;
 
-import java.util.ArrayList;
 import java.util.Collection;
-import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
-import java.util.Set;
 
 import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.PlayerID;
@@ -18,27 +14,18 @@ import games.strategy.engine.data.Unit;
  */
 public abstract class DependentBattle extends AbstractBattle {
   private static final long serialVersionUID = 9119442509652443015L;
-  protected Map<Territory, Collection<Unit>> m_attackingFromMap = new HashMap<>();
-  protected Set<Territory> m_attackingFrom = new HashSet<>();
-  protected final Collection<Territory> m_amphibiousAttackFrom = new ArrayList<>();
 
   DependentBattle(final Territory battleSite, final PlayerID attacker, final BattleTracker battleTracker,
       final boolean isBombingRun, final BattleType battleType, final GameData data) {
     super(battleSite, attacker, battleTracker, isBombingRun, battleType, data);
   }
 
-  public Collection<Territory> getAttackingFrom() {
-    return m_attackingFrom;
-  }
+  public abstract Collection<Territory> getAttackingFrom();
 
-  public Map<Territory, Collection<Unit>> getAttackingFromMap() {
-    return m_attackingFromMap;
-  }
+  public abstract Map<Territory, Collection<Unit>> getAttackingFromMap();
 
   /**
    * @return territories where there are amphibious attacks
    */
-  public Collection<Territory> getAmphibiousAttackTerritories() {
-    return m_amphibiousAttackFrom;
-  }
+  public abstract Collection<Territory> getAmphibiousAttackTerritories();
 }
diff --git a/src/main/java/games/strategy/triplea/delegate/MustFightBattle.java b/src/main/java/games/strategy/triplea/delegate/MustFightBattle.java
index 8ee7cb2..9d0bb67 100644
--- a/src/main/java/games/strategy/triplea/delegate/MustFightBattle.java
+++ b/src/main/java/games/strategy/triplea/delegate/MustFightBattle.java
@@ -64,7 +64,10 @@ public class MustFightBattle extends DependentBattle implements BattleStepString
 
   private static final long serialVersionUID = 5879502298361231540L;
   // maps Territory-> units (stores a collection of who is attacking from where, needed for undoing moves)
+  private Map<Territory, Collection<Unit>> m_attackingFromMap = new HashMap<>();
   private final Collection<Unit> m_attackingWaitingToDie = new ArrayList<>();
+  private Set<Territory> m_attackingFrom = new HashSet<>();
+  private final Collection<Territory> m_amphibiousAttackFrom = new ArrayList<>();
   private final Collection<Unit> m_defendingWaitingToDie = new ArrayList<>();
   // keep track of all the units that die in the battle to show in the history window
   private final Collection<Unit> m_killed = new ArrayList<>();
@@ -2138,6 +2141,14 @@ public class MustFightBattle extends DependentBattle implements BattleStepString
     return games.strategy.triplea.Properties.getTransportCasualtiesRestricted(m_data);
   }
 
+  /**
+   * @return territories where there are amphibious attacks
+   */
+  @Override
+  public Collection<Territory> getAmphibiousAttackTerritories() {
+    return m_amphibiousAttackFrom;
+  }
+
   private void fireOffensiveAAGuns() {
     m_stack.push(new FireAA(false));
   }
@@ -2722,6 +2733,16 @@ public class MustFightBattle extends DependentBattle implements BattleStepString
     return m2 - m1;
   }
 
+  @Override
+  public Collection<Territory> getAttackingFrom() {
+    return m_attackingFrom;
+  }
+
+  @Override
+  public Map<Territory, Collection<Unit>> getAttackingFromMap() {
+    return m_attackingFromMap;
+  }
+
   // used for setting stuff when we make a scrambling battle when there was no previous battle there, and we need
   // retreat spaces
   public void setAttackingFromAndMap(final Map<Territory, Collection<Unit>> attackingFromMap) {
diff --git a/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java b/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java
index d9bb032..7868ae9 100644
--- a/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java
+++ b/src/main/java/games/strategy/triplea/delegate/NonFightingBattle.java
@@ -33,6 +33,10 @@ import games.strategy.util.Util;
  */
 public class NonFightingBattle extends DependentBattle {
   private static final long serialVersionUID = -1699534010648145123L;
+  private final Set<Territory> m_attackingFrom = new HashSet<>();
+  private final Collection<Territory> m_amphibiousAttackFrom = new ArrayList<>();
+  // maps Territory-> units (stores a collection of who is attacking from where, needed for undoing moves)
+  private final Map<Territory, Collection<Unit>> m_attackingFromMap = new HashMap<>();
 
   public NonFightingBattle(final Territory battleSite, final PlayerID attacker, final BattleTracker battleTracker,
       final GameData data) {
@@ -94,6 +98,24 @@ public class NonFightingBattle extends DependentBattle {
     m_isOver = true;
   }
 
+  /**
+   * @return territories where there are amphibious attacks
+   */
+  @Override
+  public Collection<Territory> getAmphibiousAttackTerritories() {
+    return m_amphibiousAttackFrom;
+  }
+
+  @Override
+  public Collection<Territory> getAttackingFrom() {
+    return m_attackingFrom;
+  }
+
+  @Override
+  public Map<Territory, Collection<Unit>> getAttackingFromMap() {
+    return m_attackingFromMap;
+  }
+
   boolean hasAttackingUnits() {
     final CompositeMatch<Unit> attackingLand = new CompositeMatchAnd<>();
     attackingLand.add(Matches.alliedUnit(m_attacker, m_data));
```

With this change, I was able to open `triplea_39594_Jap5.tsvg`, start a new local game, and the `NonFightingBattle` scramble to Sea Zone 39 ran without an NPE.

You probably want to get an opinion from @ron-murhammer or @DanVanAtta on what's the best way to proceed here.

**Comment by RoiEXLab** [comment]
Related to #814 
Serialization is not calling the constructor (for obvious reasons)

**Comment by simon33-2** [comment]
That third choice represents a half way move towards move four: Revert #1583 

Not at all a fan of choices 3 or 4, although I thank @ssoloff for his efforts in debugging this. 1.9.1 seems clearly the best option when combined with #1599. In fact, I'm closing this PR on the assumption that would be done and I'm going to add some of the other needed changes to #1599 such as adding the old jar and updating the latest_version filename.

I think the reason you didn't reproduce the problem/fix I saw for the scramble was a migration issue.

One more thing I want to add. Is it possible to add some more users with write access such as @ssoloff ? There's a number of PRs open which should mostly be merged.

**Comment by ssoloff** [comment]
> Is it possible to add some more users with write access such as @ssoloff ?

I'm definitely the wrong person for that job. 😃  I lack even a moderate amount of TripleA domain knowledge and would probably just make a mess of things allowing changes through that I didn't fully understand.  @RoiEXLab would be a better candidate for committer, as he's been around for a while and has experience with a much broader spectrum of the code.

**Pull request closed without merging**